pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+
+// A worker that listens for new envelopes in the DB and sends them to subscribers
+// Assumes that there are many listeners - non-blocking updates are sent on buffered channels
+// and may be dropped if full
+type subscribeWorker struct {
+	ctx context.Context
+	log *zap.Logger
+
+	dbSubscription <-chan []queries.GatewayEnvelope
+	// Assumption: listeners cannot be in multiple slices
+	globalListeners     []listener
+	originatorListeners map[uint32][]listener
+	topicListeners      map[string][]listener
+}
+
+func startSubscribeWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	store *sql.DB,
+) (*subscribeWorker, error) {
+	log = log.Named("subscribeWorker")
+	q := queries.New(store)
+	pollableQuery := func(ctx context.Context, lastSeen db.VectorClock, numRows int32) ([]queries.GatewayEnvelope, db.VectorClock, error) {
+		envs, err := q.
+			SelectGatewayEnvelopes(
+				ctx,
+				*db.SetVectorClock(&queries.SelectGatewayEnvelopesParams{}, lastSeen),
+			)
+		if err != nil {
+			return nil, lastSeen, err
+		}
+		for _, env := range envs {
+			// TODO(rich) Handle out-of-order envelopes
+			lastSeen[uint32(env.OriginatorNodeID)] = uint64(env.OriginatorSequenceID)
+		}
+		return envs, lastSeen, nil
+	}
+
+	vc, err := q.SelectVectorClock(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		pollableQuery,
+		db.ToVectorClock(vc),
+		db.PollingOptions{
+			Interval: SubscribeWorkerPollTime,
+			NumRows:  subscribeWorkerPollRows,
+		},
+	)
+	dbChan, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+	worker := &subscribeWorker{
+		ctx:                 ctx,
+		log:                 log,
+		dbSubscription:      dbChan,
+		globalListeners:     make([]listener, 0),
+		originatorListeners: make(map[uint32][]listener),
+		topicListeners:      make(map[string][]listener),
+	}
+
+	go worker.start()
+
+	return worker, nil
+}
+
+func (s *subscribeWorker) start() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case new_batch := <-s.dbSubscription:
+			for _, row := range new_batch {
+				s.dispatch(&row)
+			}
+		}
+	}
+}
+
+func (s *subscribeWorker) dispatch(
+	row *queries.GatewayEnvelope,
+) {
+	bytes := row.OriginatorEnvelope
+	env := &message_api.OriginatorEnvelope{}
+	err := proto.Unmarshal(bytes, env)
+	if err != nil {
+		s.log.Error("Failed to unmarshal envelope", zap.Error(err))
+		return
+	}
+	for _, listener := range s.originatorListeners[uint32(row.OriginatorNodeID)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default: // TODO(rich) Close and clean up channel
+		}
+	}
+	for _, listener := range s.topicListeners[hex.EncodeToString(row.Topic)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+	for _, listener := range s.globalListeners {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+}
+
+func (s *subscribeWorker) addListeners(
```

Comment:
can I be a bit nitpicky here?

addListeners sounds like we are passing in listeners as parameters and adding them to the worker. The fact that we get back a channel that we should listen to makes it slightly worse.

I am the worst when it comes to naming, but I would have named it something like `createChannel` or something.

Reply:
Yes I agree, I was thinking of renaming this to `listen(requests)`